roachtest: add CompatibleClouds and Suites to TestSpec

RaduBerinde opened the pull request:
The roachtest tags have gotten very messy, with no better example than
that we run the nightly GCE suite using `tag:aws tag:default`.

This change is the first step to replacing the tags with more
structured information. We add two new fields to `TestSpec`:
 - `CompatibleClouds` is the set of clouds that the test can run on.
 - `Suites` is a set of test suites.

The fields are strongly typed and only allow specific values.

The intention is that we will (for example) run the nightly tests using
`roachtest run --cloud=XX --suite nightly`

In this change we mechanically assign the maximum set of compatible
clouds so that the nightly suites on GCE and AWS remain unchanged. We
assert that all tests have the two new fields *explicitly* set and
that the test suites match those selected by tags.

In follow-up changes we will add selection flags like `--suite`,
replace uses of tags in scripts, and remove the tags altogether.

Informs: #100605
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110687)
<!-- Reviewable:end -->

RaduBerinde commented:
I've got a bit of RSI from making this change :)

smg260 reviewed (commented):
> I've got a bit of RS

Thanks for your sacrifice.  The changes look good - just a small question.

> Reviewed 11 of 132 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-:-NenvbA3E-A6g0wknWrz:b-3r62cv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @RaduBerinde, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 318 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NenPyL_-8xEJhcLCa6B:-NenPyL_-8xEJhcLCa6C:bu7pohp) ([raw file](https://github.com/cockroachdb/cockroach/blob/488e535f29f865e276061e996a94a30258fe17a6/pkg/cmd/roachtest/registry/test_spec.go#L318)):*
> ```Go
> 
> // Suite names.
> const (
> ```

For my benefit (as a go learner) - why not use something like 

```
type Suite struct {  
  string  
}
var Nightly = Suite{"nightly"}
```

or 

```
type Suite int
const ( 
  Nightly Suite = iota
  Weekly
 ... 
)


```
in conjunction with a `map[string]Suite`  which would give us a compile time guarantee, versus runtime.
___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NenMlFo0mHoZVGpCSqY:-NenMlFo0mHoZVGpCSqZ:bmfovq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/488e535f29f865e276061e996a94a30258fe17a6/pkg/cmd/roachtest/registry/test_spec.go#L374)):*
> ```Go
> func (ss SuiteSet) AssertInitialized() {
> 	if ss.m == nil {
> 		panic("CloudSet not initialized")
> ```

`SuiteSet`


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-:-Neo3V6G2gNzSNi0xy-k:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 318 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NenPyL_-8xEJhcLCa6B:-Neo0gQu7_5TeQQUI7uC:b75word) ([raw file](https://github.com/cockroachdb/cockroach/blob/488e535f29f865e276061e996a94a30258fe17a6/pkg/cmd/roachtest/registry/test_spec.go#L318)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

For my benefit (as a go learner) - why not use something like 

```
type Suite struct {  
  string  
}
var Nightly = Suite{"nightly"}
```

or 

```
type Suite int
const ( 
  Nightly Suite = iota
  Weekly
 ... 
)


```
in conjunction with a `map[string]Suite`  which would give us a compile time guarantee, versus runtime.
</blockquote></details>

It's a judgement call..

The first one wouldn't work because the embedded field is not public; we'd do that like `type Suite string` (this is what we do for `Owner`). It would not be much different than what I have, other than having to cast a string in some cases.

The second one is usually preferable but in this case it's just more roundabout without a big benefit, IMO it's preferable that you can jump to the definition and see the string (which is what you care about, as that will need to be passed to roachtest).
___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NenMlFo0mHoZVGpCSqY:-Neo2WDv28qRP89yNJfP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/488e535f29f865e276061e996a94a30258fe17a6/pkg/cmd/roachtest/registry/test_spec.go#L374)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

`SuiteSet`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

smg260 reviewed (approved):
No mechanical changes in this PR - LGTM 

Would be nice to have one other set of eyes from test-eng.

> Reviewed 121 of 132 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-:-NeoNQ3vE0oSJcZdbXPO:buuuop8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)



<!-- Sent from Reviewable.io -->

srosenberg commented on pkg/cmd/roachtest/tests/cdc.go:
```diff
@@ -1001,11 +1001,13 @@ func skipLocalUnderArm64(cloud string) string {
+		Benchmark:        true,
+		Cluster:          r.MakeClusterSpec(4, spec.CPU(16), spec.Arch(vm.ArchAMD64)),
+		RequiresLicense:  true,
+		CompatibleClouds: registry.AllExceptAWS,
```
I am curious how `AllExceptAWS` was derived? A number of these tests can actually be run in AWS despite the fact that CI currently confines the set, largely to performance tests.

RaduBerinde replied:
Tests that didn't have the `aws` tag.

I did [ask](https://cockroachlabs.slack.com/archives/C023S0V4YEB/p1694786376678939?thread_ts=1694745643.434109&cid=C023S0V4YEB) if we would want any of the tests in the nightly (GCE) suite to not run in AWS nightlies if they can run on AWS, and the answer was no..

If we want a smaller nightly suite for AWS, it will have to be a separate suite. Though I wouldn't make it AWS specific, it could be something like "ReducedNightly".

In any case, when we start using these fields in `roachtest` I will add a flag to ignore cloud compatibility so it will always allow running a test manually if necessary.

RaduBerinde commented:
Ok to merge? I have a build with 0.1 select probability running here https://teamcity.cockroachdb.com/viewLog.html?buildId=11877054

srosenberg commented:
> Ok to merge? I have a build with 0.1 select probability running here https://teamcity.cockroachdb.com/viewLog.html?buildId=11877054

It appears the workers are deadlocked, but I am not entirely sure why. Last message from the build log,

```
[02:41:17 ]no work remaining; runWorker is bailing out...
```

We can't access the TC agent, and `test_runner.log` isn't copied under artifacts until the TC job completes. Let's wait until it times out; then, we can try to reconstruct the runner log to see what actually happened.

RaduBerinde commented:
Looks like it completed fine.

srosenberg commented:
> Looks like it completed fine.

Yep, looks good to merge. False alarm, it was the luck of the draw, e.g., `tpch_concurrency/no_streamer` took over 7 hours.

RaduBerinde commented:
TFTRs!

bors r+

renatolabs reviewed (commented):
Thanks for working on this!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-:-NeyBVifEdKkSE3V366w:b-4v1jpi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @RaduBerinde, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 317 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NeyAEHc09drUS4Sebsh:-NeyAEHc09drUS4Sebsi:b-8pwbs6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6585a961ec937a685ee1530bde1528c347673b0/pkg/cmd/roachtest/registry/test_spec.go#L317)):*
> ```Go
> }
> 
> // Suite names.
> ```

Where did this list of suites come from? Presumably there's some definition for each of them -- are we planning to document them later as they are added to TC scripts?

For instance, when I look at the list of suites, I'm not sure what's the difference between `nightly` and `roachtest`.
___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 333 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NeyAdVwC-lu0ZlOAbTQ:-NeyAdVx5IZcdBfr6A-u:b-yhbqfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6585a961ec937a685ee1530bde1528c347673b0/pkg/cmd/roachtest/registry/test_spec.go#L333)):*
> ```Go
> 	PebbleNightlyYCSBRace = "pebble_nightly_ycsb_race"
> 	Roachtest             = "roachtest"
> )
> ```

Another suite we could add at some point is `acceptance`. This is the set of tests that run locally as part of bors and is currently defined in the build script itself:

https://github.com/cockroachdb/cockroach/blob/235babd141c1c7a8291b1bc3401b9a6e2360b495/build/teamcity/cockroach/ci/tests/local_roachtest_impl.sh#L24


<!-- Sent from Reviewable.io -->

renatolabs commented:
Canceling since it failed on bors:

```
[18:42:34][Run local roachtests] panic: CloudSet not initialized
```

bors r-

RaduBerinde reviewed (commented):
Oops, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-:-NeydO_KB6ZqKC2V85OL:b-u2kpac)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 317 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NeyAEHc09drUS4Sebsh:-NeycxaR7_cD9TEbj4ep:b-jash54) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6585a961ec937a685ee1530bde1528c347673b0/pkg/cmd/roachtest/registry/test_spec.go#L317)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Where did this list of suites come from? Presumably there's some definition for each of them -- are we planning to document them later as they are added to TC scripts?

For instance, when I look at the list of suites, I'm not sure what's the difference between `nightly` and `roachtest`.
</blockquote></details>

The list comes from the tags. Those were not documented either, I don't know what to tell you.. Many of these are not used in any scripts that I could find..  I think we should separately try to document them and trim them down. 

'roachtest' seems to be used for some tests for roachtest itself
___
*[`pkg/cmd/roachtest/registry/test_spec.go` line 333 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110687#-NeyAdVwC-lu0ZlOAbTQ:-NeydLGgDNJpr_TQHKi6:b-n2ewn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f6585a961ec937a685ee1530bde1528c347673b0/pkg/cmd/roachtest/registry/test_spec.go#L333)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Another suite we could add at some point is `acceptance`. This is the set of tests that run locally as part of bors and is currently defined in the build script itself:

https://github.com/cockroachdb/cockroach/blob/235babd141c1c7a8291b1bc3401b9a6e2360b495/build/teamcity/cockroach/ci/tests/local_roachtest_impl.sh#L24
</blockquote></details>

Agreed.. can you file an issue?


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I think the failure was due to a recently added test: https://github.com/cockroachdb/cockroach/pull/110602

RaduBerinde committed:
roachtest: add CompatibleClouds and Suites to TestSpec

The roachtest tags have gotten very messy, with no better example than
that we run the nightly GCE suite using `tag:aws tag:default`.

This change is the first step to replacing the tags with more
structured information. We add two new fields to `TestSpec`:
 - `CompatibleClouds` is the set of clouds that the test can run on.
 - `Suites` is a set of test suites.

The fields are strongly typed and only allow specific values.

The intention is that we will (for example) run the nightly tests using
`roachtest run --cloud=XX --suite nightly`

In this change we mechanically assign the maximum set of compatible
clouds so that the nightly suites on GCE and AWS remain unchanged. We
assert that all tests have the two new fields *explicitly* set and
that the test suites match those selected by tags.

In follow-up changes we will add selection flags like `--suite`,
replace uses of tags in scripts, and remove the tags altogether.

Informs: #100605
Release note: None

RaduBerinde commented:
bors r+
